server/topic_proxy.go

```diff
@@ -30,34 +30,15 @@ func (t *Topic) runProxy(hub *Hub) {
 					log.Println("proxy topic: route join request from proxy to master failed:", err)
 				}
 			}
-
-		case leave := <-t.unreg:
-			// Detach session from topic; session may continue to function.
-			var asUid types.Uid
-			if leave.pkt != nil {
-				asUid = types.ParseUserId(leave.pkt.AsUser)
+			if join.sess.inflightReqs != nil {
+				join.sess.inflightReqs.Done()
 			}
 
-			// FIXME: The old comment is probably not true: Explicitly specify user ID because the proxy session
-			// hosts multiple client sessions.
-			if asUid.IsZero() {
-				if pssd, ok := t.sessions[leave.sess]; ok {
-					asUid = pssd.uid
-				} else {
-					log.Println("proxy topic: leave request sent for unknown session")
-					continue
-				}
-			}
-			// Remove the session from the topic without waiting for a response from the master node
-			// because by the time the response arrives this session may be already gone from the session store
-			// and we won't be able to find and remove it by its sid.
-			t.remSession(leave.sess, asUid)
-			if err := globals.cluster.routeToTopicMaster(ProxyReqLeave, leave.pkt, t.name, leave.sess); err != nil {
-				log.Println("proxy topic: route broadcast request from proxy to master failed:", err)
-			}
-			if len(t.sessions) == 0 {
-				// No more sessions attached. Start the countdown.
-				killTimer.Reset(idleProxyTopicTimeout)
+		case leave := <-t.unreg:
+			t.handleProxyLeaveRequest(leave, killTimer)
+			if leave.pkt != nil && leave.sess.inflightReqs != nil {
+				// If it's a client initiated request.
+				leave.sess.inflightReqs.Done()
```

Comment:
I think it should be `Add(-1)` here. This is a multiplexing session. The leave is for a single user within the session. Also I think `handleProxyLeaveRequest` should return success/fail or the result of `t.remSession(leave.sess, asUid)`.

Reply:
`Add(-1)` is equivalent to `Done()`.
Modified `t.handleProxyLeaveRequest()`.

Reply:
Good point. I was under a wrong impression that `Done()` sets lock count to 0.

I don't see how `leave.sess.inflightReqs` can be zero.